File: examples/typescript/part-3/08-sandboxing.ts
// Illustrative examples for notes/part-3-processing-security/08-sandboxing.md
// Not for production use. Licensed under MIT (see LICENSE-CODE).
/// <reference path="../types.d.ts" />

import { execFileSync } from "node:child_process";

interface CommandPolicy {
  allowedBinaries: Record<string, boolean>;
  timeout?: number;
  maxOutputBytes?: number;
  workDir?: string;
}

function runSandboxedCommand(
  policy: CommandPolicy,
  name: string,
  ...args: string[]
): string {
  if (!policy.allowedBinaries[name]) {
    throw new Error("binary is not allowed");
  }

  const timeoutMs = (policy.timeout && policy.timeout > 0 ? policy.timeout : 5) * 1000;
  const maxOutput = policy.maxOutputBytes && policy.maxOutputBytes > 0
    ? policy.maxOutputBytes
    : 64 * 1024;

  // Важно: execFileSync(name, args) без shell.
  // Не делать: execSync("sh -c ...user input...")
  const stdout = execFileSync(name, args, {
    cwd: policy.workDir,
    timeout: timeoutMs,
    maxBuffer: maxOutput,
    env: {
      PATH: "/usr/bin:/bin",
      LANG: "C",
    },
    encoding: "utf8",
  });

  return stdout;
}

// --- Sandbox policy для tool ---

interface SandboxPolicy {
  enabled: boolean;
  networkAllowed: boolean;
  allowedHosts?: string[];
  readOnly: boolean;
  maxInputBytes: number;
  maxOutputBytes: number;
  maxDurationSec: number;
  requiresApproval?: boolean;
}

interface ToolSpec {
  name: string;
  action: string;
  risk: string;
  sandboxPolicy: SandboxPolicy;
}

const TOOLS: Record<string, ToolSpec> = {
  read_docs: {
    name: "read_docs",
    action: "read",
    risk: "medium",
    sandboxPolicy: {
      enabled: true,
      networkAllowed: false,
      readOnly: true,
      maxInputBytes: 1 << 20,
      maxOutputBytes: 64 << 10,
      maxDurationSec: 5,
    },
  },
  run_code: {
    name: "run_code",
    action: "execute",
    risk: "high",
    sandboxPolicy: {
      enabled: true,
      networkAllowed: false,
      readOnly: false,
      maxInputBytes: 128 << 10,
      maxOutputBytes: 64 << 10,
      maxDurationSec: 3,
      requiresApproval: true,
    },
  },
};

/** EVAL-NETWORK-PREFLIGHT-01: env/network flags before start_agent. */
interface NetworkPreflightState {
  dnsPublic?: boolean;
  ipv4Public?: boolean;
  ipv6Public?: boolean;
  httpProxyPresent?: boolean;
  httpsProxyPresent?: boolean;
  allProxyPresent?: boolean;
  cloudMetadataReachable?: boolean;
  localhostPrivilegedServices?: boolean;
}

/** True → do not start agent. */
function failsNetworkPreflight(s: NetworkPreflightState): boolean {
  return Boolean(
    s.dnsPublic ||
      s.ipv4Public ||
      s.ipv6Public ||
      s.httpProxyPresent ||
      s.httpsProxyPresent ||
      s.allProxyPresent ||
      s.cloudMetadataReachable ||
      s.localhostPrivilegedServices,
  );
}

export { runSandboxedCommand, TOOLS, failsNetworkPreflight };

export type { CommandPolicy, SandboxPolicy, ToolSpec, NetworkPreflightState };

File: examples/python/part-3/08-sandboxing.py
# Illustrative examples for notes/part-3-processing-security/08-sandboxing.md
# Not for production use. Licensed under MIT (see LICENSE-CODE).

from __future__ import annotations

import subprocess
from dataclasses import dataclass, field


@dataclass
class CommandPolicy:
    allowed_binaries: dict[str, bool]
    timeout: float = 5.0
    max_output_bytes: int = 64 * 1024
    work_dir: str = ""


def run_sandboxed_command(
    policy: CommandPolicy,
    name: str,
    *args: str,
) -> str:
    if not policy.allowed_binaries.get(name):
        raise ValueError("binary is not allowed")

    timeout = policy.timeout if policy.timeout > 0 else 5.0
    max_output = policy.max_output_bytes if policy.max_output_bytes > 0 else 64 * 1024

    # Важно: subprocess.run([name, *args], shell=False).
    # Не делать: subprocess.run("sh -c ...user input...", shell=True)
    result = subprocess.run(
        [name, *args],
        capture_output=True,
        text=True,
        timeout=timeout,
        cwd=policy.work_dir or None,
        env={
            "PATH": "/usr/bin:/bin",
            "LANG": "C",
        },
        check=False,
    )

    if result.returncode != 0:
        raise RuntimeError(result.stderr or "command failed")

    stdout = result.stdout or ""
    if len(stdout.encode("utf-8")) > max_output:
        stdout = stdout.encode("utf-8")[:max_output].decode("utf-8", errors="ignore")

    return stdout


# --- Sandbox policy для tool ---


@dataclass
class SandboxPolicy:
    enabled: bool = False
    network_allowed: bool = False
    allowed_hosts: list[str] = field(default_factory=list)
    read_only: bool = False
    max_input_bytes: int = 0
    max_output_bytes: int = 0
    max_duration_sec: int = 0
    requires_approval: bool = False


@dataclass
class ToolSpec:
    name: str
    action: str
    risk: str
    sandbox_policy: SandboxPolicy


TOOLS: dict[str, ToolSpec] = {
    "read_docs": ToolSpec(
        name="read_docs",
        action="read",
        risk="medium",
        sandbox_policy=SandboxPolicy(
            enabled=True,
            network_allowed=False,
            read_only=True,
            max_input_bytes=1 << 20,
            max_output_bytes=64 << 10,
            max_duration_sec=5,
        ),
    ),
    "run_code": ToolSpec(
        name="run_code",
        action="execute",
        risk="high",
        sandbox_policy=SandboxPolicy(
            enabled=True,
            network_allowed=False,
            read_only=False,
            max_input_bytes=128 << 10,
            max_output_bytes=64 << 10,
            max_duration_sec=3,
            requires_approval=True,
        ),
    ),
}
